### PR TITLE
Address nullable feedback for System.Runtime and System.Runtime.Extensions

### PR DESCRIPTION
**Pull request by safern**
I added the changes to TextWriter taking a null buffer to follow pattern and start discussion in the PR. If not agreed, I can always revert. Said that, I expect corefx tests to fail, will disable them and update them.

cc: @dotnet/nullablefc @jkotas

### CODE REVIEW

File: src/System.Private.CoreLib/shared/System/AppDomain.cs [hunk 1]
```diff
@@ -101,9 +101,9 @@ public static AppDomain CreateDomain(string friendlyName)
             throw new PlatformNotSupportedException(SR.PlatformNotSupported_AppDomains);
         }
 
-        public int ExecuteAssembly(string assemblyFile) => ExecuteAssembly(assemblyFile, null);
+        public int ExecuteAssembly(string assemblyFile) => ExecuteAssembly(assemblyFile, null!);
 
-        public int ExecuteAssembly(string assemblyFile, string?[]? args)
+        public int ExecuteAssembly(string assemblyFile, string?[] args)
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/AppDomain.cs]: Fix the `ExecuteAssembly` overload to match

Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/AppDomain.cs]: Why are we disallowing null here when it is a valid input? It does not make sense to me.

Comment (safern) [reply src/System.Private.CoreLib/shared/System/AppDomain.cs]: Actually, I should've done it in the overload that takes a params. My bad, good catch.

Comment (safern) [reply src/System.Private.CoreLib/shared/System/AppDomain.cs]: Sorry it was my mistake. I wanted to do it in the one that takes params. I miss read.

Comment (safern) [reply src/System.Private.CoreLib/shared/System/AppDomain.cs]: We agreed that we have a pattern in the framework that APIs that take a params shouldn't allow null.

File: src/System.Private.CoreLib/shared/System/IO/TextWriter.cs [hunk 2]
```diff
@@ -595,8 +596,13 @@ public Task WriteAsync(char[]? buffer)
             return WriteAsync(buffer, 0, buffer.Length);
         }
 
-        public virtual Task WriteAsync(char[] buffer, int index, int count)
+        public virtual Task WriteAsync(char[]? buffer, int index, int count)
         {
+            if (buffer == null)
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/IO/TextWriter.cs]: This needs to check arguments. If buffer is null, index and count better be null as well.

Comment (safern) [reply src/System.Private.CoreLib/shared/System/IO/TextWriter.cs]: Right, will fix.

**Comment by jkotas** [comment]
> I added the changes to TextWriter taking a null buffer to follow pattern

Note that the contract for these method is documented in the documentation:
https://docs.microsoft.com/en-us/dotnet/api/system.io.textwriter.write?view=netcore-3.0#System_IO_TextWriter_Write_System_Char___System_Int32_System_Int32_
And that these methods are virtual - folks who overwrite them will be still implementing the old contract.

I do not think this change is an improvement. It is likely introducing more inconsistencies than it is fixing.

File: src/System.Private.CoreLib/shared/System/AppDomain.cs (hunk 1 above)
Comment (safern) [reply src/System.Private.CoreLib/shared/System/AppDomain.cs]: Now that I give it a second thought I don't think that is correct in this case. We're even calling our own overrides with null here. Will leave as nullable.

**Comment by jkotas** [comment]
Also, similar null checks are in majority of the APIs that take (array, index, length). Is there *any* existing API that takes (array, index, length) and allows array to be null when the length is null?

**Comment by safern** [comment]
> Is there any existing API that takes (array, index, length) and allows array to be null when the length is null?

I just searched all of corelib and actually just found 2 instances of it:

https://source.dot.net/#System.Private.CoreLib/shared/System/Text/StringBuilder.cs,771
and
https://source.dot.net/#System.Private.CoreLib/src/System/Utf8String.Construction.cs,138

that when the type of the array is a `char` didn't search for other types, but yeah, I think if it is documented this way, I don't think is worth changing the implementation in this case, specially since all these are virtual.

File: src/System.Private.CoreLib/shared/System/AppDomain.cs (hunk 1 above)
Comment (safern) [reply src/System.Private.CoreLib/shared/System/AppDomain.cs]: Fixed

**Comment by safern** [comment]
I reverted the change to AppDomain and TextWriter/StreamWriter.

**Pull request merged**